doc/decisions/DR024-Versioning-traits-and-specifications-method.md

```diff
@@ -0,0 +1,412 @@
+
+
+### Host side
+```python
+# The host side implementation is almost entirely unchanged from
+# from standard, excepting a new BatchElementError type
+trait_set = {LocatableContentTrait.kId}
+context = self.createTestContext()
+
+result = [None]
+
+def success_cb(idx, traits_data):
+    result[0] = traits_data
+
+def error_cb(idx, batchElementError):
+    if instanceOf(batchElementError, VersionNotSupportedError):
+      incompatible_with_manager_error()
+    ...
+
+self._manager.resolve(
+    [entity_reference], trait_set, ResolveAccess.kRead, context, success_cb, error_cb
+)
+```
+### Manager side
+```python
+# On the manager side, each trait gets unpacked (probably via a utility
+# method provided by traitgen or openassetio), and branching is done
+# on a trait by trait basis
+
+def resolve(
+        self,
+        entityReferences,
+        traitSet,
+        resolveAccess,
+        context,
+        hostSession,
+        successCallback,
+        errorCallback,
+    ):
+
+  # This function is imagined a pre-provided utility, possibly generated
+  # via traitgen, that would error in some way if the trait set does not
+  # contain the locatableContent trait.
+  locatableContent_version = locatableContent_version_from_trait_set(traitSet)
+
+  success_result = TraitsData()
+  if (locatableContent_version == 2):
+    trait = mediacreation.traits.content.LocatableContentTrait(success_result)
+    trait.setUrl(managed_filesystem_locations[ref.toString()])
+    successCallback(idx, success_result)
+  elif (locatableContent_version == 1):
+    trait = mediacreation.legacy.traits.content.LocatableContentTrait_v1(success_result)
+    trait.setLocation(managed_filesystem_locations[ref.toString()])
+    successCallback(idx, success_result)
+  else:
+    raise RuntimeError()
+
+```
+
+#### Pros
+- More flexible, traits of different versions can be mixed in the same
+  session, which could allow hosts/managers to perform partial upgrades
+  in the case they are blocked on small specifics, and allow components
+  of a system to each work with a different schema version and merge
+  their results.
+- Hosts don't have to interrogate and declare which library versions
+  are in use, which helps especially in the case of second-order
+  dependencies.
+- Seems to fit with existing OpenAssetIO design, especially with
+  `BatchElementError`, as an incompatible version wouldn't necessarily
+  need to error the whole batch.
+- No new API, core `openassetio` library remains version unaware.
+- Cannot be ignored, unlike other "layered on top" mechanisms.
+- Managers not required to consider storage/communication of additional
+  version data to enable interpretation by future downstream systems.
+- Prior art (e.g. OTIO) encodes versions in their data structure,
+  presumably for good reason.
+
+#### Cons
+- No "primary" version across the session, increasing implementation
+  complexity.
+- Introduces "Try and see" negotiation on the host side, although
```

Comment:
[question] Is this strictly true as the existing `managementPolicy`methods would , without change, allow introspection of support?

Reply:
I believe you would still need to do a try-and-see implementation pattern though, in that you'd need to provide a set of all possible versions to `managementPolicy` and have it tell you which ones it likes.

However, i'll grant I wasn't thinking of that when I wrote this con, and that approach is not really any less maintainable than a "ask for what versions are good" approach, at least from a hosts perspective. Making `managementPolicy` even _more_ complicated for the manager though ... eesh.

What do you think @feltech? I am tempted to remove this as a con based on this observation.

Reply:
And I suppose it surfaces a potential con that management policy is meant to be entity agnostic (as its about what the  manager can support), and that different entities in a simple manager may not have the same version. But I guess it has always been the case that not every entity would have any specific trait though even though "it should have", so maybe its no different to the situation were already in... as 'simple' managers usually respond more generally ie "yes to everything".

Reply:
Changed this up a bit, simplified just to call out the complexity, and link to the relevant section. Removed the try-and-see thing as you're correct it's not strictly true. Think that works.

https://github.com/OpenAssetIO/OpenAssetIO/pull/1267/commits/066101d9e574e43e101aea7e2590ddf0527b2c3d